pkg/audit/manager.go

```diff
@@ -0,0 +1,270 @@
+				oname := item.GetName()
+				oapiVersion := item.GetAPIVersion()
+				ogvk := item.GroupVersionKind()
+				okind := ogvk.Kind
+				ogroup := ogvk.Group
+				oversion := ogvk.Version
+				violations, found, err := unstructured.NestedSlice(item.Object, "status", "violations")
+				if err != nil {
+					return err
+				}
+				if !found || len(violations) == 0 {
+					fmt.Printf("constraint %s violations is either not found or empty, skip clear \n", oname)
+				} else {
+					fmt.Printf("constraint name %s, apiversion %s, kind %s, group %s, version %s, violations %v \n", oname, oapiVersion, okind, ogroup, oversion, violations)
+					var emptyAuditResults []auditResult
+					err = updateConstraintStatus(dynamic, ogroup, oversion, okind, oname, oapiVersion, emptyAuditResults)
+					if err != nil {
+						return err
+					}
+					fmt.Printf("constraint %s violations have been cleared \n", oname)
+				}
+			}
+		}
+		// update each violated constraint
+		for selfLink := range updateList {
+			auditResults := updateList[selfLink]
+			aResult := auditResults[0]
+			return updateConstraintStatus(dynamic, aResult.cgroup, aResult.cversion, aResult.ckind, aResult.cname, aResult.capiversion, auditResults)
+		}
+	}
+	return nil
+}
+
+func (am *auditManager) auditManagerLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Audit Manager close")
+			close(am.stopper)
+			return
+		default:
+			time.Sleep(time.Duration(*auditInterval) * time.Second)
+			if err := am.audit(); err != nil {
+				log.Error(err, "audit manager audit() failed")
+			}
+		}
+	}
+}
+
+func startMgr(mgr manager.Manager, stopper chan struct{}, stopped chan<- struct{}) {
+	if err := mgr.Start(stopper); err != nil {
+		log.Error(err, "Error starting audit watch manager")
+	}
+	// mgr.Start() only returns after the manager has completely stopped
+	close(stopped)
+	log.Info("audit exiting")
+}
+
+func updateConstraintStatus(dynamic dynamic.Interface, cgroup string, cversion string, ckind string, cname string, capiversion string, auditResults []auditResult) error {
+	cresource := fmt.Sprintf("%ss", strings.ToLower(ckind))
+	cstrClient := dynamic.Resource(schema.GroupVersionResource{Group: cgroup, Version: cversion, Resource: cresource})
+	o, err := cstrClient.Get(cname, metav1.GetOptions{TypeMeta: metav1.TypeMeta{Kind: ckind, APIVersion: capiversion}})
+	if err != nil {
+		return err
+	}
+	val, found, err := unstructured.NestedBool(o.Object, "status", "enforced")
+	if err != nil {
+		return err
+	}
+	if !found {
+		return errors.Errorf("constraint %s status enforced not found", cname)
+	}
+	if !val {
+		return errors.Errorf("constraint %s status not enforced", cname)
+	}
+	// create constraint status violations
+	var statusViolations []interface{}
+	for _, ar := range auditResults {
+		statusViolations = append(statusViolations, map[string]string{
+			"kind":      ar.rkind,
+			"name":      ar.rname,
+			"namespace": ar.rnamespace,
```

Comment:
Namespace should probably be optional, to conform with standards for cluster-scoped resources.

I think this code could be compacted by having a statusViolation object that has a `json:"namespace,omitempty"` tag.